Design/spectrum_number_and_workspace_index_abstraction.md

```diff
@@ -0,0 +1,212 @@
+- For better compatibility for the current interface and for users who do not like to deal with spectrum numbers it should also be possible to specify workspace indices.
+  It is important to note that this is a **global** workspace index.
+  As explained above, in an MPI-based Mantid run this is **not** the same as the workspace index.
+- Some algorithms also accept detector IDs and then identify corresponding spectrum numbers.
+
+For developers there should be single property that provides all this:
+
+- When declaring the property it must be possible to specify which of the index types the algorithm should accept.
+- A reasonable default would be *spectrum number* and *global workspace index*.
+- *Detector ID* is used in few cases and should be optional.
+- The word *global* should most likely be omitted from the user interface.
+  The translation from global to local workspace indices happens between the UI and the algorithm, i.e., neither the user nor (generally) the developer needs to know about global indices.
+  Putting *global* in the UI would only create confusion.
+
+#### DetectorIDsProperty
+
+Some algorithms work directly with detectors and thus need a detector index.
+This is similar to the `SpectrumNumbersProperty` but simpler.
+Most likely it would be sufficient to provide only *detector ID* as possible input.
+
+It should be pointed out that:
+
+- Algorithms that work with *spectra* use `SpectrumNumbersProperty` to translate from a user provided *detector ID* to an *index for a spectrum* (in a workspace).
+- Algorithms that work with *detectors* use `DetectorIDsProperty` to translate from a user provided *detector ID* to an *index for a detector* (in an instrument).
+
+
+## Internals
+
+#### New functionality
+
+- Introduce a class that deals with translation and validation of indices.
+  This will replace code that is now in `MatrixWorkspace` (translation) and in many algorithms (validation).
+  Let us call this class `Translator` for the rest of this document.
+  - In contrast to the current situation this can be unit tested very well.
+  - Defines a consistent way of dealing with errors in the index specification.
+    This is basically also part of a consistent user interface.
+  - Reduces amount of code in algorithms and makes them more readable.
+- Provide a convenience function or method for declaring all relevant algorithm properties in a single statement.
+  - Reduces amount of code in algorithms and makes them more readable.
+  - Automatically gives consistent interface.
+- Spectrum numbers can be defined differently in each workspace.
+  As a consequence the object dealing with translation must be linked to a workspace.
+  For example, it could be a member of the workspace.
+- We should make sure that we do not negatively influence performance when a trivial set of indices is specified by the user.
```

Comment:
I don't think our performance tests are in general sensitive enough for this. Probably we will be fine if we support some special cases (which are actually the most common ones):

- We want to avoid going through a `map` or `vector` for translation in the simple cases of "full-range" (all spectra), or a simple "min-max" case. As you saw, I already implemented this in the prototyping I did for the `Indexing` library. I believe this will cover 95% of cases and ensure that there will be no performance hit.

- In the other cases, when an actual list of spectra is given, I suppose that that list is commonly small (since it is often manual user input), so `map` access is negligible.

- The remaining performance critical case is loading event data. Each event has an associated detector ID, and this needs to be translated to a workspace index. This is currently done by creating a `vector` (instead of a `map`) for look-up. We certainly have to support a similar mechanism. This is a case where we will need to write some performance tests, to prevent regression.